bpf/lib/mcast.h

```diff
@@ -0,0 +1,440 @@
+	__be32 saddr;
+	/* local ifindex of subscriber of exit interface is remote subscriber */
+	__u32 ifindex;
+	/* reserved */
+	__u16 pad1;
+	/* reserved */
+	__u8  pad2;
+	/* flags for further subscriber description */
+	__u8  flags;
+};
+
+#ifdef ENABLE_SYNTHETIC_MULTICAST
+
+#define MCAST_MAX_GROUP 1024
+#define MCAST_MAX_SUBSCRIBERS 1024
+/* used to bound iteration of group records within an igmpv3 membership report */
+#define MCAST_MAX_GREC 24
+
+/* Multicast group map is a nested hash of maps.
+ * The outer map is keyed by a 'mcast_group_v4' multicast group address.
+ * The inner value is an hash map of 'mcast_subscriber_v4' structures keyed
+ * by a their IPv4 source address in big endian format.
+ */
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH_OF_MAPS);
+	__type(key, mcast_group_v4);
+	__type(value, __u32);
+	__uint(pinning, LIBBPF_PIN_BY_NAME);
+	__uint(max_entries, MCAST_MAX_GROUP);
+	__uint(map_flags, CONDITIONAL_PREALLOC);
+	/* Multicast group subscribers inner map definition */
+	__array(values, struct {
+		__uint(type, BPF_MAP_TYPE_HASH);
+		__uint(key_size, sizeof(__be32));
+		__uint(value_size, sizeof(struct mcast_subscriber_v4));
+		__uint(max_entries, MCAST_MAX_SUBSCRIBERS);
+		__uint(map_flags, CONDITIONAL_PREALLOC);
+	});
+} MCAST_GROUP_OUTER_V4_MAP __section_maps_btf;
+
+/* lookup a subscriber map for the given ipv4 multicast group
+ * returns a void pointer to a inner subscriper map if one exists
+ */
+static __always_inline void *mcast_lookup_subscriber_map(__be32 *group)
+{
+	return map_lookup_elem(&MCAST_GROUP_OUTER_V4_MAP, group);
+}
+
+/* returns 1 if ip4 header is followed by an IGMP payload, 0 if not */
+static __always_inline bool mcast_ipv4_is_igmp(const struct iphdr *ip4)
+{
+	if (ip4->protocol == IPPROTO_IGMP)
+		return 1;
+	return 0;
+}
+
+/* returns the IGMP type for a given IGMP message
+ * a call to 'mcast_ipv4_is_igmp' must be used prior to this call to ensure an
+ * igmp message follows the ipv4 header
+ */
+static __always_inline __s32 mcast_ipv4_igmp_type(const struct iphdr *ip4,
+						  const void *data,
+						  const void *data_end)
+{
+	const struct igmphdr *hdr;
+	int ip_len = ip4->ihl * 4;
+
+	if (data + ETH_HLEN + ip_len + sizeof(struct igmphdr) > data_end)
+		return DROP_INVALID;
+
+	hdr = data + ETH_HLEN + ip_len;
+	return hdr->type;
+}
+
+/* add a subscriber to a subscriber map */
+/* returns 1 on success or DROP_INVALID for error */
+static __always_inline __s32 mcast_ipv4_add_subscriber(void *map,
+						       struct mcast_subscriber_v4 *sub)
+{
+	if ((map_update_elem(map, &sub->saddr, sub, BPF_ANY) != 0))
+		return DROP_INVALID;
+	return 1;
+}
+
+/* remove a subscriber to a subscriber map */
+/* always returns 1 */
+static __always_inline void mcast_ipv4_remove_subscriber(void *map,
+							 struct mcast_subscriber_v4 *sub)
+{
+	map_delete_elem(map, &sub->saddr);
+}
+
+static __always_inline __s32 mcast_ipv4_handle_v3_membership_report(void *ctx,
```

Comment:
This style of function declaration takes up quite a bit of vertical space, is this what checkpatch expects? :thinking:

Reply:
If IIRC this style was one that checkpatch stopped barking at me for.

Reply:
Resolving for now, if you do feel strongly just re-open and ping me.